torch_xla/csrc/aten_xla_type.cpp

```diff
@@ -1643,11 +1649,19 @@ at::Tensor& XLANativeFunctions::index_put_(
   XLA_CHECK(self.scalar_type() == values.scalar_type());
   CanonicalIndexInfo canonical_index_info =
       GetCanonicalIndexInfo(self, indices);
-  XLATensor self_tensor = bridge::GetXlaTensor(self);
+  c10::optional<Device> device =
+      bridge::GetXlaDevice(canonical_index_info.base);
+  if (!device.has_value()) {
```

Comment:
Why don't we apply this `if/else` logic in `GetCAnonicalIndexInfo()`? Is that because we need `device` for `bridge::GetOrCreateXlaTensor(values, *device)`? If yes, the next question is, why do we believe an upstream `value` tensor "can" be on a different device?

(My understanding is `base` and `index` are the root cause of this failure; correct me if I am wrong.)

Reply:
I put it here mostly because we also handle the binary op tensor [device mismatch](https://github.com/pytorch/xla/blob/master/torch_xla/csrc/aten_xla_type.cpp#L115) in this file. 

Index is of type `c10::List<c10::optional<at::Tensor>>` which is something `GetOrCreateXlaTensor` can't handle. We generally not trying to change the `at::tensor` device type but create the `XLATensor` on the correct device. `at::tensor` -> `XLATensor` happens in this file.

Reply:
Got it. Given it's used in three spots, why don't we do this logic as a separate function call?

Reply:
It is a bit awkward to add a helper function like

```
bridge::GetXlaDevice(at::Tensor tensor, vector<at::Tensor> tensors);
```
one cleaner way of doing this is to append the single tensor to the end of the vector and use the vector version of the `GetXlaDevice`, but that felt like an overkill since we either need to create a new vector or manually pop the `base` after this call.

Reply:
Yeah, maybe this implementation is the most appropriate of all.